RavenDB-17243: Implement native support for nulls

redknightlois opened the pull request:
After some working around I was able to piggyback the null support in an space efficient way and with as less as possible disruption into the format. More complex format means that the decoding phase would be more costly to execute. Therefore, keeping the format as simple as possible was a requirement. 

Current implementation supports handling null at every property level (albeit some options are not implemented yet on this draft PR). 

cc @maciejaszyk check if there is something else that's required to use native nulls instead of null strings.

redknightlois committed:
RavenDB-17243: Support for asking for null values.

redknightlois committed:
RavenDB-17243: Implementation of skippable null bit stream.

ayende commented on src/Corax/IndexEntryBuilder.cs:
```diff
@@ -256,14 +229,41 @@ public void FinishWritingEnumerable(int metadataLocation, int dataLocation, int
+            // Write the pointer to the location
             MemoryMarshal.Write(stringPtrTableLocation, ref dataLocation);
+
+            // We will include null values if there are nulls to be stored.           
+            int nullBitStreamSize = values.Length / (sizeof(byte) * 8) + values.Length % (sizeof(byte) * 8) == 0 ? 0 : 1;
+            byte* nullStream = stackalloc byte[nullBitStreamSize];
+            bool hasNull = false;
+            for (int i = 0; i < values.Length; i++)
+            {
+                if (values.IsNull(i))
+                {
+                    hasNull = true;
+                    PtrBitVector.SetBitInPointer(nullStream, i, true);
+                }
+            }
+
+            if (hasNull)
+            {
+                // Copy the null stream.
+                new ReadOnlySpan<byte>(nullStream, nullBitStreamSize * sizeof(long))
+                    .CopyTo(_buffer.Slice(dataLocation));
+
+                dataLocation += nullBitStreamSize;
+
+                // Signal that we will have to deal with the nulls.
+                _buffer[indexEntryFieldLocation] |= (byte)IndexEntryFieldType.HasNull;
+            }
```
This is duplicated code, no?

ayende commented on src/Corax/IndexEntryBuilder.cs:
```diff
@@ -305,7 +308,34 @@ public void Write(int field, IReadOnlySpanEnumerator values, ReadOnlySpan<long>
 
+            // Write the pointer to the location.
             MemoryMarshal.Write(stringPtrTableLocation, ref dataLocation);
+
+            // We will include null values if there are nulls to be stored.           
+            int nullBitStreamSize = values.Length / (sizeof(byte) * 8) + values.Length % (sizeof(byte) * 8) == 0 ? 0 : 1;
+            byte* nullStream = stackalloc byte[nullBitStreamSize];
+            bool hasNull = false;
+            for (int i = 0; i < values.Length; i++)
+            {
+                if (values.IsNull(i))
+                {
+                    hasNull = true;
+                    PtrBitVector.SetBitInPointer(nullStream, i, true);
+                }
+            }
+
+            if (hasNull)
+            {
+                // Copy the null stream.
+                new ReadOnlySpan<byte>(nullStream, nullBitStreamSize * sizeof(long))
+                    .CopyTo(_buffer.Slice(dataLocation));
+
+                dataLocation += nullBitStreamSize;
+
+                // Signal that we will have to deal with the nulls.
+                _buffer[indexEntryFieldLocation] |= (byte)IndexEntryFieldType.HasNull;
+            }
```
Duplicated code

redknightlois replied:
Yes, for now it will stay like that.

ayende commented on src/Corax/IndexEntryBuilder.cs:
```diff
@@ -473,51 +504,85 @@ internal IndexEntryFieldIterator(IndexEntryFieldType type)
+                if (!Type.HasFlag(IndexEntryFieldType.HasNull))
+                    return false;
+
+                throw new NotImplementedException();
```
??

redknightlois replied:
Still not implemented... I use that to be able to kill the cases where I use that feature :)

ayende commented on src/Corax/IndexEntryBuilder.cs:
```diff
@@ -842,12 +916,28 @@ public bool Read(int field, out IndexEntryFieldType type, out Span<byte> value,
+                        byte* nullTablePtr = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(_buffer));
+                        if (PtrBitVector.GetBitInPointer(nullTablePtr + spanTableOffset, elementIdx) == true)
+                            goto HasNull;
+                    }
```
No need to skip the null table here?

redknightlois replied:
Null table does not exist if there are no potential nulls stored. So no need to skip. This is a pay-for-usage type of thing... No NULL, no pay.

ayende commented on src/Corax/IndexEntryBuilder.cs:
```diff
@@ -842,12 +916,28 @@ public bool Read(int field, out IndexEntryFieldType type, out Span<byte> value,
+                        byte* nullTablePtr = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(_buffer));
+                        if (PtrBitVector.GetBitInPointer(nullTablePtr + spanTableOffset, elementIdx) == true)
+                            goto HasNull;
+                    }
```
No need to skip the null table here?

ayende commented on test/FastTests/Corax/IndexEntryWriter.cs:
```diff
@@ -31,7 +31,15 @@ public StringArrayIterator(string[] values)
-            public ReadOnlySpan<byte> this[int i] => Encoding.UTF8.GetBytes(_values[i]);
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
```
Do we need it, the `_values` already does this, no?

ayende commented on test/FastTests/Corax/IndexSearcher.cs:
```diff
@@ -52,6 +52,14 @@ public StringArrayIterator(IEnumerable<string> values)
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
+
```
Do we need it, the `_values` already does this, no?

ayende commented on test/FastTests/Sparrow/Encoder3GramTests.cs:
```diff
@@ -33,6 +33,14 @@ private struct StringKeys : IReadOnlySpanEnumerator, ISpanEnumerator
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
```
Do we need it, the `_values` already does this, no?

redknightlois replied:
Not really

ayende commented on test/Tryouts/CoraxEnron.cs:
```diff
@@ -301,6 +301,14 @@ public StringArrayIterator(IEnumerable<string> values)
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
```
Do we need it, the `_values` already does this, no?

ayende commented on test/Tryouts/Program.cs:
```diff
@@ -90,6 +90,14 @@ public StringArrayIterator(string[] values)
 
+            public bool IsNull(int i)
+            {
+                if (i < 0 || i >= Length)
+                    throw new ArgumentOutOfRangeException();
+
```
Do we need it, the `_values` already does this, no?

ayende reviewed (commented):
If I understand correctly, null handling is using:

* An additional flag on the entries (`HasNull`)
* If that flag is set, we add a bitfield that mark null values for the particular fields?

How would you then handle a scenario an array with null in it?

`{ Stages: [ "stages/1", null, "stages/3"] }`

And then asking for: `Stages = null` ?

redknightlois commented:
> If I understand correctly, null handling is using:
> 
> * An additional flag on the entries (`HasNull`)
> * If that flag is set, we add a bitfield that mark null values for the particular fields?
> 
> How would you then handle a scenario an array with null in it?
> 
> `{ Stages: [ "stages/1", null, "stages/3"] }`
> 
> And then asking for: `Stages = null` ?

There are 2 different ways to access that. The `Stages = null` actually returns null for the item entirely. It will be written differently than the case of  `{ Stages: [ "stages/1", null, "stages/3"] }` . At the protocol level, these are entirely different formats. One is a list, the other is a plain null on the `Stages` field.

ayende commented:
> 

`Stages = null` means - currently - a `Stages` field hat has a `null` value in it. Not just set to null, but also an array that has a null value.

redknightlois commented:
> > 
> 
> `Stages = null` means - currently - a `Stages` field hat has a `null` value in it. Not just set to null, but also an array that has a null value.

Yes. Those are two different. The case where 'Stages' is 'null' has the flag set up as null. Flags is equal to 0. The case of the array containing a null value is a HasNull and with the internal marker for nulls.

redknightlois committed:
RavenDB-17243: Added null vector after sequence stream.

redknightlois committed:
RavenDB-17243: Support for high level nulls.

redknightlois committed:
RavenDB-17243: Refactor

arekpalinski commented:
test this please

redknightlois committed:
RavenDB-17602: Index Entry Field Type changed to ushort

redknightlois commented:
The change to use `ushort` instead of `byte` as type header is done. Required standardizing the access.

arekpalinski commented:
@redknightlois can you check `SlowTests.Corax.RavenDB_18377.SlowTests.Corax.RavenDB_18377.ArrayInIndex(options: DatabaseMode = Single , SearchEngineMode = Corax)` test?

redknightlois committed:
RavenDB-17602: Support for empty lists

redknightlois committed:
RavenDB-17243: Fix inverted clause in test.

redknightlois commented:
@arekpalinski Either we fixed (inadvertedly) a bug when working on NULL support OR it was already wrong. The conditions were inverted between the static index and the auto index. There is absolutely no reason why they would be returning Empty and the Single and both pass in green.

arekpalinski commented:
It was a corax specific test so I believe we just put the wrong conditions which was passing at that time, even though they were not valid.

redknightlois committed:
RavenDB-17243: Indexing process would treat nulls as real nulls.

redknightlois committed:
RavenDB-17243: Fixed the case of null fields appearing first.

redknightlois committed:
RavenDB-17243: Support for querying explicit nulls as a term

redknightlois commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexWriteOperation.cs:
```diff
@@ -66,9 +66,6 @@ public override void Commit(IndexingStatsScope stats)
-                return;
             
             using (Stats.AddStats.Start())
                 _indexWriter.Index(lowerId, data, _knownFields);
```
@maciejaszyk Is there any way I can know here or before if the document is actually an implicit null (and/or explicit)?

maciejaszyk replied:
@redknightlois 
In `CoraxDocumentConverterBase` we have something in constructor:
![image](https://user-images.githubusercontent.com/86351904/168743666-266973e3-37bd-48ed-9d7c-7a058b8c7c05.png)

redknightlois committed:
RavenDB-17243: Include 10s of tests to use Corax.

redknightlois commented:
I am enabling tests to be run on Corax with the objective of finding edge cases. So far, found 1.

arekpalinski commented:
Seems like there is just one failure - https://jenkins.hibernatingrhinos.com/job/PR_Tests/1903/testReport/junit/SlowTests.Bugs.Queries/QueryingOnEqualToNull/SlowTests_Bugs_Queries_QueryingOnEqualToNull_QueryingOnEqNotNull_options___DatabaseMode___Single___SearchEngineMode___Corax_/

redknightlois committed:
RavenDB-17243: When the parameter is an slice, the analyzer is not been applied.

redknightlois commented:
@maciejaszyk Can you look into the `QueryingOnEqNotNull` failing test? It seems to me that it is really unrelated to #14158 and more related to case sensitivity and how we treat the `id()` field. Missing some context to fix it. If you need we can have a call in my morning. 

@arekpalinski If all the slow tests pass, coordinate with Maciej as this PR is a dependency for #14164.

redknightlois committed:
RavenDB-17243: Nulls need to be treated differently depending on usage.

Reimplement with null support the general cases and built an optimized version for Null and NotNull.

arekpalinski commented:
test this please

maciejaszyk commented on src/Corax/Queries/UnaryMatch.Ordinal.cs:
```diff
@@ -191,129 +435,147 @@ private static int FillFuncNumerical<TComparer>(ref UnaryMatch<TInner, TValueTyp
+            UnaryMatchOperationMode mode = UnaryMatchOperationMode.Any, int take = -1)
         {
             if (typeof(TValueType) == typeof(Slice))
             {
                 return new UnaryMatch<TInner, TValueType>(
                     in inner, UnaryMatchOperation.NotEquals,
                     searcher, fieldId, value,
-                    &FillFuncSequence<NotEqualsMatchComparer>, &AndWith,
-                    inner.Count, inner.Confidence.Min(QueryCountConfidence.Normal), distinct, take);
+                    mode == UnaryMatchOperationMode.Any ? &FillFuncSequenceAny<EqualsMatchComparer> : &FillFuncSequenceAll<EqualsMatchComparer>, &AndWith,
+                    inner.Count, inner.Confidence.Min(QueryCountConfidence.Normal), mode, true, take: take);
             }
             else
             {
                 return new UnaryMatch<TInner, TValueType>(
                     in inner, UnaryMatchOperation.NotEquals,
                     searcher, fieldId, value,
-                    &FillFuncNumerical<NotEqualsMatchComparer>, &AndWith,
-                    inner.Count, inner.Confidence.Min(QueryCountConfidence.Normal), distinct, take);
+                    mode == UnaryMatchOperationMode.Any ? &FillFuncNumerical<EqualsMatchComparer> : &FillFuncNumerical<EqualsMatchComparer>, &AndWith,
+                    inner.Count, inner.Confidence.Min(QueryCountConfidence.Normal), mode, true, take: take);
```
You are passing `EqualsMatchComparer` in `NotEqualMatch` ;) 
@redknightlois

maciejaszyk replied:
This is about `QueryingOnEqNotNull`

redknightlois replied:
I am passing `EqualsMatchComparer` but I am also forcing `distinct = true` ... this is one of the side effects of `null` you have to be very careful with the `Any` and `All` distinction. It is not symmetric as it was before... maybe the failed ones are related to `FillFuncNumerical` as I am seeing that they have no null handling.

maciejaszyk commented:
Some other tests failed after the fix - https://github.com/ravendb/ravendb/pull/14244
Probably this is related to UnaryMatch logic.

maciejaszyk commented:
BTW: 
I think we should skip the `null` value at `ExistsMatch`

redknightlois commented:
@maciejaszyk Check `QueryingOnEqNotNull` it seems it is not even related to NULLs it just surfaced because of it. The problem (at least it seems to me) is one of casing of `id()` fields.

redknightlois committed:
RavenDB-17243: Lowering of ids.

redknightlois committed:
RavenDB-17243: Handling the edge case of nulls in NotEquals
